tensorflow_quantum/core/serialize/serializer_test.py

```diff
@@ -425,15 +473,17 @@ class SerializerTest(tf.test.TestCase, parameterized.TestCase):
 
     @parameterized.parameters([{
         'circ_proto_pair': v
-    } for v in _get_valid_circuit_proto_pairs()])
+    } for v in _get_controlled_circuit_proto_pairs() +
+                               _get_circuit_proto_pairs()])
     def test_serialize_circuit_valid(self, circ_proto_pair):
         """Test conversion of cirq Circuits to tfq_gate_set proto."""
         self.assertProtoEquals(serializer.serialize_circuit(circ_proto_pair[0]),
                                circ_proto_pair[1])
 
     @parameterized.parameters([{
         'circ_proto_pair': v
-    } for v in _get_valid_circuit_proto_pairs()])
+    } for v in _get_controlled_circuit_proto_pairs() +
+                               _get_circuit_proto_pairs()])
```

Comment:
I'm having trouble understanding why this test is passing.  For example, the first entry in the list returned by `_get_circuit_proto_pairs()` is an HPow gate.  `_make_controlled_circuit` should promote this to a `cirq.ControlledOperation`, and the corresponding proto is promoted to have `control_qubits` and `control_values` fields.  But deserializing that proto should yield a `DelayedAssignmentGate`; how does it come out equal to the `cirq.ControlledOperation` in `circ_proto_pair[0]`?

Reply:
The DelayedAssignmentGate is what is getting sent into the `op_serializer.py` functions. These functions will very soon call the `.on` method which the `DelayedAssignmentGate` uses to create an instance of the apropriate gate for serialization. Without it all `cirq.ControlledOperations` look the same to the serializer and parsing them would be more of a pain.

Reply:
Yeah the DelayedAssignmentGate is needed to separate out the controlling qubits.  I just mean in the test below there are parameters such that `serializer.deserialize_circuit(circ_proto_pair[1])` should be a `DelayedAssignmentGate` while `circ_proto_pair[0]` should be a `cirq.ControlledOperation`, and I was wondering how these evaluate to equal

Reply:
Here is the life of a proto as it gets deserialized:

Proto -> `op_deserializer.py` -> begins looping over operations and invokes the specific deserialization rule for the current operation -> that procedure returns a gate that has not yet had qubit's chosen via a call to the `.on` method (An instance of delayedassignmentgate) -> `op_deserializer.py` calls the `.on` method and the DelayedAssignmentGate returns a `cirq.ControlledOperation` with the `sub_gate` equal to the original gate found in the proto. 

This gets called only within `op_deserializer.py`:
```python3
def on(self, *qubits):
        """Returns gate_callable on qubits controlled by contol_qubits."""
        return self._gate_callable(*qubits).controlled_by(
            *self._control_qubits, control_values=self._control_values)
```

So the DelayedAssignmentGate only exists for a moment before returning the controlledoperation that is the true form of the gate. The `DelayedAssignmentGate` class is needed so that we can assign control qubit information outside of `op_deserializer.py` since it doesn't natively support controlled gates.



Does this help clear things up ?

Reply:
Thanks for the details here, I understand what's happening now: the `DelayedAssignmentGate` only exists between lines 105 and 106 of `cirq.google.op_deserializer`, after that it is the regular `ControlledOperation`; so the new class is a workaround for the existing op `from_proto` implementation.

Could you add a short addition to the DelayedAssignmentGate to clarify things for future readers, maybe something like "DelayedAssignmentGate is only ever instantiated locally by the native Cirq deserializer (inside cirq.google.GateOpDeserializer.from_proto) and then discarded"?

Reply:
Added more clarity in the comments found in the `serialize_circuit` function.